File: gwo_np.cpp
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <time.h>
#include <math.h>

struct Feature {
    int story_points;
    int satisfaction;
};

struct Wolf {
    double fitness;
    double *position;
};

int size = 10;

Feature problem[10] = {
        {3,  20},
        {5,  50},
        {2,  40},
        {13, 100},
        {5,  40},
        {8,  10},
        {5,  20},
        {5,  40},
        {3,  30},
        {3,  10}};

// Limit of story points to consider for selection
int max_story_points = 20;

double gen(double min, double max) {
    double range = (max - min);
    double div = RAND_MAX / range;
    return min + (rand() / div);
}

void imprime_vet(double v[], int tam) {
    int i;
    for (i = 0; i < tam; i++)
        printf("%.6f ", v[i]);
    printf("\n");
}

int fibbonacci(int n) {
    if (n == 0) {
        return 0;
    } else if (n == 1) {
        return 1;
    } else {
        return (fibbonacci(n - 1) + fibbonacci(n - 2));
    }
}

void merge(Wolf arr[], int l, int m, int r) {
    int i, j, k;
    int n1 = m - l + 1;
    int n2 = r - m;

    Wolf L[n1], R[n2];

    for (i = 0; i < n1; i++)
        L[i] = arr[l + i];
    for (j = 0; j < n2; j++)
        R[j] = arr[m + 1 + j];

    i = 0;
    j = 0;
    k = l;
    while (i < n1 && j < n2) {
        if (L[i].fitness <= R[j].fitness) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

void mergeSort(Wolf arr[], int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;

        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);

        merge(arr, l, m, r);
    }
}

double fitness(double *position, int dim) {
    int fitness_value = 0;
    int total_story_points = 0;
    for (int i = 0; i < dim; i++) {
        if (position[i] >= 1) {
            total_story_points += problem[i].story_points;
            fitness_value += problem[i].satisfaction;
        }
    }

    if (total_story_points > max_story_points || total_story_points == 0) {
        return HUGE_VAL;
    } else {
        return 100.0 / (fitness_value * (total_story_points * 0.1));
    }
}

void wolf(Wolf *w, int dim, double minx, double maxx) {
    for (int i = 0; i < dim; i++) {
        w->position[i] = gen(minx, maxx);
    }

    w->fitness = fitness(w->position, dim);
}

double calculate_a(bool explore, int iter, int max_iter, int explore_counter) {
    if (explore == false) {
        return 2 * (1 - iter / max_iter);
    } else {
        return 2 * (1 + iter / max_iter) * (gen(0.0, 10.0) * explore_counter);
    }
}

double *gwo(int max_iter, int n, int dim, int minx, int maxx) {

    Wolf *population = (Wolf *) malloc(sizeof(Wolf) * n);

    for (int i = 0; i < n; i++) {
        population[i].position = new double[dim];

        wolf(&population[i], dim, minx, maxx);
    }

    mergeSort(population, 0, n - 1);

    Wolf *alpha_wolf = &population[0];
    Wolf *beta_wolf = &population[1];
    Wolf *gamma_wolf = &population[2];

    bool explore = false;

    int iter = 0;

    // Maximum value of the fitness function
    int previous_alpha = 1;
    // Expands the search space at each cycle of iterations if not the best solution was found in the last cycle of iterations
    // A cycle is composed of 10 iterations
    int explore_counter = 1;
    double max_iter_multiplier = 1.0;

    while (iter < (max_iter * max_iter_multiplier) && (max_iter_multiplier < 2)) {

        if (iter % 10 == 0 && iter > 1) {

            printf("Iter = %d best fitness = %f\n", iter, alpha_wolf->fitness);
            if (alpha_wolf->fitness == previous_alpha) {
                explore = true;
                explore_counter++;
            } else {
                explore = false;
                explore_counter = 1;
                max_iter_multiplier = max_iter_multiplier + 0.025;
            }
        }

        double a = calculate_a(explore, iter, max_iter, explore_counter);

        for (int i = 0; i < n; i++) {
            double X1[dim];
            double X2[dim];
            double X3[dim];
            double nextPos[dim];

            for (int j = 0; j < dim; j++) {
                double A1 = a * (2 * gen(0.0, 1.0) - 1);
                double A2 = a * (2 * gen(0.0, 1.0) - 1);
                double A3 = a * (2 * gen(0.0, 1.0) - 1);

                double C1 = 2 * gen(0.0, 1.0);
                double C2 = 2 * gen(0.0, 1.0);
                double C3 = 2 * gen(0.0, 1.0);

                X1[j] = alpha_wolf->position[j] - A1 * abs(
                        C1 * alpha_wolf->position[j] - population[i].position[j]);

                X2[j] = beta_wolf->position[j] - A2 * abs(
                        C2 * beta_wolf->position[j] - population[i].position[j]);

                X3[j] = gamma_wolf->position[j] - A3 * abs(
                        C3 * gamma_wolf->position[j] - population[i].position[j]);

                nextPos[j] = X1[j] + X2[j] + X3[j];
            }

            for (int k = 0; k < dim; k++) {
                nextPos[k] = nextPos[k] / dim + 1;
            }

            double nextFitness = fitness(nextPos, dim);

            if (nextFitness < population[i].fitness) {
                for (int k = 0; k < dim; k++) {
                    population[i].position[k] = nextPos[k];
                }
                population[i].fitness = nextFitness;
            }
        }

        mergeSort(population, 0, n - 1);

        previous_alpha = alpha_wolf->fitness;

        alpha_wolf = &population[0];
        beta_wolf = &population[1];
        gamma_wolf = &population[2];

        iter++;
    }
    return alpha_wolf->position;
}

int main() {

    srand((unsigned) time(NULL));
    /*
    int size = (rand() % 10) * 5;
    Feature *problem = (Feature *) malloc(sizeof(Feature) * size);
    for (int i = 0; i < size; i++) {
      problem[i].story_points = fibbonacci(rand() % 10);
      problem[i].satisfaction = rand() % 100;
    }
    */

    printf("\nBegin grey wolf optimization on NRP client satisfaction\n");

    int dim = size;

    printf("Goal is to minimize %d variables\n", dim);

    int num_particles = 75;
    int max_iter = size * size;

    printf("Setting num_particles = %d\n", num_particles);

    printf("Setting max_iter = %d\n", (max_iter));
    printf("\nStarting GWO algorithm\n");

    double *best_position = gwo(max_iter, num_particles, dim, 0.0, 2.0);

    printf("\nBest solution found:\n");

    printf("[");
    int total_sp = 0;
    int total_satisfaction = 0;
    bool after_first = false;
    for (int i = 0; i < dim; i++) {
        if (best_position[i] > 1) {
            if (after_first) {
                printf(", ");
            } else {
                after_first = true;
            }
            printf("F%d", i);

            total_sp = total_sp + problem[i].story_points;
            total_satisfaction = total_satisfaction + problem[i].satisfaction;
        }
    }
    printf("]\n");

    printf("Total story points: %d\n", total_sp);
    printf("Total satisfaction: %d\n", total_satisfaction);

    imprime_vet(best_position, dim);

    double err = fitness(best_position, dim);

    printf("Fitness of best solution = %.6f\n", err);

    return 0;
}

File: gwo_np.lua
math.randomseed((math.tan(os.time())) * ((os.clock() * 1000) + (os.time() / 1000)))

function shallowcopy(orig)
    local orig_type = type(orig)
    local copy
    if orig_type == 'table' then
        copy = {}
        for orig_key, orig_value in pairs(orig) do
            copy[orig_key] = orig_value
        end
    else
        copy = orig
    end
    return copy
end

function fibonacci(n)
  local function inner(m, a, b)
    if m == 0 then
      return a
    end
    return inner(m-1, b, a+b)
  end
  return inner(n, 0, 1)
end

function quickSort(array, p, r)
    p = p or 1
    r = r or #array
    if p < r then
        q = partition(array, p, r)
        quickSort(array, p, q - 1)
        quickSort(array, q + 1, r)
    end
end

function partition(array, p, r)
    local x = array[r].fitness
    local i = p - 1
    for j = p, r - 1 do
        if array[j].fitness <= x then
            i = i + 1
            local temp = array[i]
            array[i] = array[j]
            array[j] = temp
        end
    end
    local temp = array[i + 1]
    array[i + 1] = array[r]
    array[r] = temp
    return i + 1
end

local fitness_client_satisfaction = function (max_story_points, problem)
  return function (position)
    local fitness_value = 0.0
    local total_story_points = 0.0
    for i = 1,#position do
      if (position[i] > 1) then
        total_story_points = total_story_points + problem[i].story_points
        fitness_value = fitness_value + problem[i].satisfaction
      end
    end

    if total_story_points > max_story_points  then
      return math.huge
    else
      return 1/fitness_value
    end
  end
end

local wolf = function (fitness, dim, minx, maxx)
  local self = {}

  self.position = {}

  for i = 1,dim do
    self.position[i] = ((maxx - minx) * math.random() + minx)
  end

  self.fitness = fitness(self.position)

  return self
end

local gwo = function(fitness, max_iter, n, dim, minx, maxx)

  local population = {}

  for i = 1, n do
    population[i] = wolf(fitness, dim, minx, maxx)
  end

  quickSort(population, 1, #population)

  local alpha_wolf, beta_wolf, gamma_wolf = shallowcopy(population[1]), shallowcopy(population[2]), shallowcopy(population[3])

  local explore = false

  local iter = 0
  local alpha_const = 2
  local previous_alpha = 1
  local explore_counter = 1
  local max_iter_multiplier = 1.0
  while (iter < (max_iter * max_iter_multiplier) and (max_iter_multiplier < 2)) do

    if (iter % 10 == 0 and iter > 1) then

      print(string.format("Iter = " .. iter .. " best fitness = %.6f", alpha_wolf.fitness))
      if (string.format("%.6f",alpha_wolf.fitness) ==  previous_alpha) then
        explore = true
        explore_counter = explore_counter + 1
      else
        explore = false
        explore_counter = 1
        max_iter_multiplier = max_iter_multiplier + 0.025
      end
    end

    local a = (function()
        if (explore == false) then
          return alpha_const *(1 - iter/max_iter);
        else
          return alpha_const *(1 + iter/max_iter) * (math.random()*explore_counter);
        end;
      end)()
    print(a)

    for i = 1, n do
      local X1 = {}
      local X2 = {}
      local X3 = {}
      local nextPos = {}

      local A1, A2, A3 = a * (alpha_const * math.random() - 1), a * (
              alpha_const * math.random() - 1), a * (alpha_const * math.random() - 1)
      local C1, C2, C3 = alpha_const * math.random(), alpha_const*math.random(), alpha_const*math.random()

      for j = 1,dim do

        X1[j] = alpha_wolf.position[j] - A1 * math.abs(
        C1 * alpha_wolf.position[j] - population[i].position[j])

        X2[j] = beta_wolf.position[j] - A2 * math.abs(
        C2 *  beta_wolf.position[j] - population[i].position[j])

        X3[j] = gamma_wolf.position[j] - A3 * math.abs(
        C3 * gamma_wolf.position[j] - population[i].position[j])

        nextPos[j] = X1[j] + X2[j] + X3[j]
      end

      for k  = 1,dim do
        nextPos[k] = nextPos[k] / dim
      end

      local nextFitness = fitness(nextPos)

      if nextFitness < population[i].fitness then
          population[i].position = nextPos
          population[i].fitness = nextFitness
      end

    end
    quickSort(population, 1, #population)

    previous_alpha = string.format("%.6f",alpha_wolf.fitness)

    alpha_wolf, beta_wolf, gamma_wolf = shallowcopy(population[1]), shallowcopy(population[2]), shallowcopy(population[3])

    iter = iter + 1
  end

  return alpha_wolf.position
end

--[[
local problem = {}
for i = 1, math.random(1,10) * 5 do
  problem[i] = {}
  problem[i].story_points = fibonacci(math.random(1,10))
  problem[i].satisfaction = math.random(1,10) * 10
end
]]--

local problem = {
    {story_points = 3,satisfaction = 20},
    {story_points = 5,satisfaction = 50},
    {story_points = 2,satisfaction = 40},
    {story_points = 13,satisfaction = 100},
    {story_points = 5, satisfaction = 40},
    {story_points = 8, satisfaction = 10},
    {story_points = 5,satisfaction = 20},
    {story_points = 5,satisfaction = 40},
    {story_points = 3,satisfaction = 30},
    {story_points = 3,satisfaction = 10}
}

print("\nBegin grey wolf optimization on NRP client satisfaction\n")
local dim = #problem
local fitness = fitness_client_satisfaction(20, problem)

print("Goal is to minimize  " .. dim .. " variables")

local num_particles = 75
local max_iter = math.pow(#problem,2)

print("Setting num_particles = " .. num_particles)
print("Setting max_iter    = " .. (max_iter))
print("\nStarting GWO algorithm\n")

local best_position = gwo(fitness, max_iter, num_particles, dim, 0.0, 2.0)

print("\nBest solution found:")

io.write("[")
local total_sp = 0
local total_satisfaction = 0
local after_first = false
for i = 1,dim do
  if (best_position[i] > 1) then
    if (after_first) then
      io.write(", ")
    else
      after_first = true
    end
    io.write('F'..i)

    total_sp = total_sp + problem[i].story_points
    total_satisfaction = total_satisfaction + problem[i].satisfaction
  end
end
print("]")

print("Total story points: " .. total_sp)
print("Total satisfaction: " .. total_satisfaction)

local err = fitness(best_position)

print(string.format("Fitness of best solution = %.6f", err))